src/crl.rs

```diff
@@ -52,6 +52,86 @@ pub struct CertRevocationList<'a> {
     pub crl_number: Option<&'a [u8]>,
 }
 
+/// Representation of a RFC 5280[^1] profile Certificate Revocation List (CRL) revoked certificate
+/// entry.
+///
+/// [^1]: <https://www.rfc-editor.org/rfc/rfc5280#section-5>
+pub struct RevokedCert<'a> {
+    /// Serial number of the revoked certificate.
+    pub serial_number: &'a [u8],
+
+    /// The date at which the CA processed the revocation.
+    pub revocation_date: Time,
+
+    /// Identifies the reason for the certificate revocation.
+    pub reason_code: Option<RevocationReason>,
+
+    /// Provides the date on which it is known or suspected that the private key was compromised or
+    /// that the certificate otherwise became invalid. This date may be earlier than the revocation
+    /// date which is the date at which the CA processed the revocation.
+    pub invalidity_date: Option<Time>,
+}
+
+/// Identifies the reason a certificate was revoked.
+/// See RFC 5280 §5.3.1[^1]
+///
+/// [^1] <https://www.rfc-editor.org/rfc/rfc5280#section-5.3.1>
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[allow(missing_docs)] // Not much to add above the code name.
+pub enum RevocationReason {
+    /// Unspecified should not be used, and is instead assumed by the absence of a RevocationReason
+    /// extension.
+    Unspecified,
+    KeyCompromise,
+    CaCompromise,
+    AffiliationChanged,
+    Superseded,
+    CessationOfOperation,
+    CertificateHold,
+    /// RemoveFromCrl only appears in delta CRLs that are unsupported.
+    RemoveFromCrl,
+    PrivilegeWithdrawn,
+    AaCompromise,
+}
+
+#[derive(Clone, Copy)]
+/// Controls the behaviour of [iter_revoked_certs].
+pub enum RevokedCertIteration<T> {
+    /// Continue iterating revoked certificates.
+    KeepGoing,
+
+    /// Stop iteration of revoked certificates, yielding a result.
+    Stop(Result<T, Error>),
+}
+
+impl<'a> CertRevocationList<'a> {
+    /// Iterate through the certificate revocation list's revoked certificates (if any).
+    /// For each revoked certificate the provided callback is invoked with a [RevokedCert] and
+    /// allowed to return a [RevokedCertIteration] indicating whether iteration should proceed. If
+    /// no call to f returns [RevokedCertIteration::Stop], or if there were no revoked certificates
+    /// in the CRL, the [result_if_never_stopped_early] result is returned.
+    pub fn iter_revoked_certs<T>(
```

Comment:
FWIW I think this pattern is extremely obtuse but I haven't dug in enough to figure out if it can be replaced by something more idiomatic without losing some key guarantee.

Reply:
That's fair. I think in this case at least it's easy enough to do what we're looking for with `IntoIterator` and a helper. I've reworked using that approach.